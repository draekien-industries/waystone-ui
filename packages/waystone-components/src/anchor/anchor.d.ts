/** @jsxImportSource theme-ui */
import { FC } from 'react';
import { LinkProps } from 'theme-ui';
/** The props for the anchor component. */
export interface AnchorProps extends LinkProps {
    /**
     * A flag indicating whether the href is an external link.
     * @default false
     */
    external?: boolean;
}
export declare const Anchor: FC<AnchorProps>;
//# sourceMappingURL=anchor.d.ts.map