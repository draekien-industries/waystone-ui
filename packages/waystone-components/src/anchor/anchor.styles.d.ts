import { ThemeUIStyleObject } from 'theme-ui';
export declare const anchorCss: ThemeUIStyleObject;
//# sourceMappingURL=anchor.styles.d.ts.map