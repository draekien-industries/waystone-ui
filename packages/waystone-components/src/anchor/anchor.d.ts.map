{"version":3,"file":"anchor.d.ts","sourceRoot":"","sources":["anchor.tsx"],"names":[],"mappings":"AAAA,gCAAgC;AAChC,OAAO,EAAE,EAAE,EAAE,MAAM,OAAO,CAAC;AAC3B,OAAO,EAAQ,SAAS,EAAE,MAAM,UAAU,CAAC;AAG3C,0CAA0C;AAC1C,MAAM,WAAW,WAAY,SAAQ,SAAS;IAC5C;;;OAGG;IACH,QAAQ,CAAC,EAAE,OAAO,CAAC;CACpB;AAqBD,eAAO,MAAM,MAAM,EAAE,EAAE,CAAC,WAAW,CAMlC,CAAC"}