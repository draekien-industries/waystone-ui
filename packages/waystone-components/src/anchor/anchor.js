var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx } from "theme-ui/jsx-runtime";
import { Link } from 'theme-ui';
import { anchorCss } from './anchor.styles';
/** The internal anchor component */
const InternalAnchor = (_a) => {
    var { children } = _a, rest = __rest(_a, ["children"]);
    return (_jsx(Link, Object.assign({}, rest, { sx: anchorCss }, { children: children })));
};
/** The external anchor component */
const ExternalAnchor = (_a) => {
    var { children, target = '_blank', rel = 'noopener noreferrer' } = _a, rest = __rest(_a, ["children", "target", "rel"]);
    return (_jsx(Link, Object.assign({}, rest, { sx: anchorCss, target: target, rel: rel }, { children: children })));
};
export const Anchor = (_a) => {
    var { external = false } = _a, rest = __rest(_a, ["external"]);
    if (external) {
        return _jsx(ExternalAnchor, Object.assign({}, rest));
    }
    return _jsx(InternalAnchor, Object.assign({}, rest));
};
