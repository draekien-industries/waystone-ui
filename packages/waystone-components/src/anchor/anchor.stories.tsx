import React from 'react';
import { Meta, Story } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Anchor, AnchorProps } from './anchor';

export default {
  title: 'Components/Anchor',
  component: Anchor,
  args: {
    external: false,
  },
} as Meta;

const Template: Story<AnchorProps> = (args) => (
  <Anchor {...args} onClick={action('clicked')} />
);

export const Default = Template.bind({});

Default.args = {
  children: 'Anchor',
};

export const WithExternal = Template.bind({});

WithExternal.parameters = {
  docs: {
    description: {
      story:
        'Indicates the provided `href` is a link to an external resource. Automatically adds `_blank` to the `target` prop and `"noopener noreferrer"` to the `rel` prop.',
    },
  },
};

WithExternal.args = {
  href: 'https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a',
  external: true,
  children: 'Then <a> element',
};
