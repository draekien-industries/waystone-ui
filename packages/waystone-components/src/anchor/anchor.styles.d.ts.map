{"version":3,"file":"anchor.styles.d.ts","sourceRoot":"","sources":["anchor.styles.tsx"],"names":[],"mappings":"AAAA,OAAO,EAAE,kBAAkB,EAAE,MAAM,UAAU,CAAC;AAE9C,eAAO,MAAM,SAAS,EAAE,kBAgBvB,CAAC"}