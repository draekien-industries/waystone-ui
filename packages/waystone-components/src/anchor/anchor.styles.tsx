import { getOutlineCss } from '@waystone/core';
import { ThemeUIStyleObject } from 'theme-ui';

const outline = getOutlineCss();

export const anchorCss: ThemeUIStyleObject = {
  paddingX: 'xs',
  textDecoration: 'none',
  borderBottom: '1px solid',
  borderBottomColor: 'highlight',
  color: 'inherit',
  transition: 'all 100ms',
  fontFamily: 'inherit',
  fontWeight: 'inherit',
  fontSize: 'inherit',
  cursor: 'pointer',
  ...outline.base,
  ':hover': {
    backgroundColor: 'ghost',
    borderTopLeftRadius: 'sm',
    borderTopRightRadius: 'sm',
  },
  ':focus': {
    backgroundColor: 'ghost',
    borderTopLeftRadius: 'sm',
    borderTopRightRadius: 'sm',
    ...outline.focused[':enabled:not(active):focus'],
  },
};

export const undecoratedAnchorCss: ThemeUIStyleObject = {
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  cursor: 'pointer',
};
