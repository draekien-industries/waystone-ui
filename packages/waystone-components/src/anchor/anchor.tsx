/** @jsxImportSource theme-ui */
import { HTMLAttributeAnchorTarget, ReactNode } from 'react';
import { Link, LinkProps } from 'theme-ui';
import { Icon } from '../icon/icon';
import { anchorCss, undecoratedAnchorCss } from './anchor.styles';

/**
 * The props for the anchor component.
 * @extends LinkProps
 */
export interface AnchorProps extends LinkProps {
  /**
   * A flag indicating whether the href is an external link.
   * @default false
   */
  external?: boolean;
  /**
   * The target to open the link in. The default value only applies when the `external` flag is `true`.
   */
  target?: HTMLAttributeAnchorTarget;
  /**
   * The relationship between the linked resource and the current document.
   * The default value only applies when the `external` flag is `true`.
   */
  rel?: string;
  /**
   * The react component to display inside the anchor.
   */
  children: ReactNode;
  /** Should the anchor render without css styles */
  noDecoration?: boolean;
}

/** The internal anchor component */
const InternalAnchor = ({
  children,
  noDecoration,
  ...rest
}: LinkProps & Pick<AnchorProps, 'noDecoration'>) => (
  <Link {...rest} sx={noDecoration ? undecoratedAnchorCss : anchorCss}>
    {children}
  </Link>
);

/** The external anchor component */
const ExternalAnchor = ({
  children,
  noDecoration,
  target = '_blank',
  rel = 'noopener noreferrer',
  ...rest
}: LinkProps & Pick<AnchorProps, 'noDecoration'>) => (
  <Link
    {...rest}
    sx={!noDecoration ? anchorCss : undecoratedAnchorCss}
    target={target}
    rel={rel}
  >
    {children} <Icon name="open_in_new" size="sm" sx={{ pb: 'xs' }} />
  </Link>
);

/**
 * A styled HTML `<a>` element which creates a hyperlink to the provided URL.
 * Content passed in via the `children` prop should indicate the link's destination.
 * See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a for more details
 * about the HTML `<a>` element.
 */
export const Anchor = ({ external = false, ...rest }: AnchorProps) => {
  if (external) {
    return <ExternalAnchor {...rest} />;
  }

  return <InternalAnchor {...rest} />;
};
