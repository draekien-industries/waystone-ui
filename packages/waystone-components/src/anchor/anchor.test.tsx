import React from 'react';
import { render } from '../common';
import { Anchor } from './anchor';

describe('The Anchor component', () => {
  it('Matches snapshot', () => {
    const { container } = render(<Anchor href="#">anchor</Anchor>);

    expect(container).toMatchSnapshot();
  });
});
