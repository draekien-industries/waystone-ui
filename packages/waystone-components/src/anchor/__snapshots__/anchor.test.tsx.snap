// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Anchor component Matches snapshot 1`] = `
<div>
  <a
    class="css-14v99eo-InternalAnchor"
    href="#"
  >
    anchor
  </a>
</div>
`;
